README.md
# Spot.IM React-Native Module 🚀

This library provides an easy integration with Spot.IM into a React-Native app.

## Requirements

* iOS 10.3 or later.
* Android SDK verison (API 19) and above.
* Have a [Spot.IM](https://spot.im) account

## Getting started

### Install react-native-spotim package

#### [@spot.im/react-native-spotim](https://www.npmjs.com/package/@spot.im/react-native-spotim)
1. Install and add the package to your project:
    `npm install @spot.im/react-native-spotim --save`
2. Import Spot.IM modules:
    `import { SpotIM, SpotIMEventEmitter, SpotIMAPI } from '@spot.im/react-native-spotim';`
3. When the app starts, call the init method with your spot-id to initilize the SDK
    `SpotIMAPI.init("spot-id");`

### Use Spot.IM native view
### iOS
1. Go to the project ios folder and make sure `use_frameworks!` is set in the Podfile
2. If you're using RN version > 0.62, you will need to disable `flipper` by doing the follownig:
  * Comment the following lines from the Podfile:
  ```ruby
  ...
  #add_flipper_pods!
  #  post_install do |installer|
  #  flipper_post_install(installer)
  #end
  ...
  ```
  * Disable flipper init by removing the following lines from AppDelegate.m:
  ```obj-c
  ...
  #if DEBUG
    #import <FlipperKit/FlipperClient.h>
    #import <FlipperKitLayoutPlugin/FlipperKitLayoutPlugin.h>
    #import <FlipperKitUserDefaultsPlugin/FKUserDefaultsPlugin.h>
    #import <FlipperKitNetworkPlugin/FlipperKitNetworkPlugin.h>
    #import <SKIOSNetworkPlugin/SKIOSNetworkAdapter.h>
    #import <FlipperKitReactPlugin/FlipperKitReactPlugin.h>

    static void InitializeFlipper(UIApplication *application) {
      FlipperClient *client = [FlipperClient sharedClient];
      SKDescriptorMapper *layoutDescriptorMapper = [[SKDescriptorMapper alloc] initWithDefaults];
      [client addPlugin:[[FlipperKitLayoutPlugin alloc] initWithRootNode:application withDescriptorMapper:layoutDescriptorMapper]];
      [client addPlugin:[[FKUserDefaultsPlugin alloc] initWithSuiteName:nil]];
      [client addPlugin:[FlipperKitReactPlugin new]];
      [client addPlugin:[[FlipperKitNetworkPlugin alloc] initWithNetworkAdapter:[SKIOSNetworkAdapter new]]];
      [client start];
    }
  #endif
  ...
  #if DEBUG
    InitializeFlipper(application);
  #endif
  ...
  ```
3. The default React-Native root view controller is an instance of UIViewController.
Spot.IM is using UINavigationController no navigate to native view controllers.
To support native navigation wrap the app with a navigation controller.

Add the following to your code:

`AppDelegate.h`:
```obj-c
@property (nonatomic, strong) UINavigationController *navControll;
```

`AppDelegate.m`:
```obj-c
...
UIViewController *rootViewController = [UIViewController new];
self.navControll = [[UINavigationController alloc] initWithRootViewController:rootViewController];
[self.navControll setNavigationBarHidden:YES]; // Hide nav bar if you don't use native navigation controller
rootViewController.view = rootView;
self.window.rootViewController = self.navControll;
...
```

### Android

- Android SDK verison (API 19) and above.
- Your application will need a permission to use the Internet. Add the following line to your **AndroidManifest.xml**:
```xml
<uses-permission android:name="android.permission.INTERNET" />
```

1. Add the following lines to your **project** module's `build.gradle` file.
```gradle
repositories {
   maven { url 'https://jitpack.io' }
}
```
2. Add the following lines to the **app** module's `build.gradle` file.
```gradle
implementation 'androidx.multidex:multidex:2.0.1'
implementation 'com.github.SpotIM.spotim-android-sdk:spotim-sdk:1.6.1'
```
3. Apply Spot.IM gradle plugin
	There are two options to implement the plugin:
	1. Using the [plugins DSL](#using-the-plugins-dsl)
	2. Using [legacy plugin application](#using-legacy-plugin-application)

	### Using the plugins DSL

	Add the following lines to the **app** module's `build.gradle` file.
	```gradle
	plugins {
	  id "im.spot" version "1.0"
	}
	```
	⚠️ **Note:** Maku sure to apply the plugin after the `com.android.application` plugin.

	### Using legacy plugin application

	1. Add the following lines to your **project** module's `build.gradle` file.
	```gradle
	buildscript {
	  repositories {
	    maven {
	      url "https://plugins.gradle.org/m2/"
	    }
	  }
	  dependencies {
	    classpath "gradle.plugin.im.spot:spotim-gradle-plugin:1.0"
	  }
	}
	```

	2. Add the following lines to the **app** module's `build.gradle` file.

	```gradle
	defaultConfig {
        ...
        multiDexEnabled true
    }
	```

	```gradle
	apply plugin: "im.spot"
	```

### Load PreConversation View in React-Native

```javascript
<SpotIM
    spotId="<SPOT.IM_ID>"
    postId="<POST_ID>"
    url="<POST_URL>"
    title="<POST_TITLE>"
    subtitle="<POST_SUBTITLE>"
    thumbnailUrl="<POST_THUMBNAIL>"
    darkModeBackgroundColor="#<HEX_COLOR>"
    style={{flex: 1}}
/>
```

#### Listen to PreConversation View size changes
We make sure the container view is resized when the PreConversation View is filled with comments.
You can also subscribe to `SpotIMEventEmitter` with `startLoginFlow` event to get PreConversation View height changes:
```javascript
SpotIMEventEmitter.addListener('viewHeightDidChange', (event) => {
    this.setState({height: event['newHeight']});
})
```

### Flows

Our SDK exposes one major flow to set up. The pre-conversation view is a view that displays a preview of 2-16 comments from the conversation, a text box to create new comments and a button to see all comments.

The Pre-conversation view should be displayed in your article view below the article.

When the user wants to see more comments we push a new ViewController/Activity which displays all comments from the conversation.

When clicking on the text box to create a new comments we bring the user to the creation screen. The users needs to be logged in inorder to post new comments, this is where the hosting app needs to integrate it's authentication system.

#### Authentication
To utilize SSO authentication, you can use your login UI by subscribing to `startLoginFlow` event:
```javascript
const onStartLoginFlow = (event) => {
    ...
}
const subscription = SpotIMEventEmitter.addListener('startLoginFlow', onStartLoginFlow);
```

##### Authentication with SSO:

There are to types of SSO available: **Generic SSO** and **SSO with JWT secret**. Please contact your Spot.IM advisor to pick the best option for you.

##### Generic SSO

1. Authenticate a user with your backend - see Authentication above
2. Call `startSSO` function and get `codeA`
3. Send the `codeA` and all needed information to your backend system in order to get `codeB`
4. Call `completeSSO` with the `codeB`
5. Check `success` and `error` properties in the callback to ensure everything is ok

###### Example
```javascript
// 2
SpotIMAPI.startSSO()
    .then(response => {
        console.log(response);
    })
    .catch(error => {
        console.error(error);
    })

// 4
SpotIMAPI.completeSSO("<CODE_B>")
    .then(response => {
        console.log(response);
    })
    .catch(error => {
        console.error(error);
    })
```


##### SSO with JWT secret

1. Authenticate a user with your backend
2. Call `sso(JWTSecret)` function with a user JWT secret
3. If there’s no error in the callback and `response?.success` is true, the authentication process finished successfully

###### Example
```javascript
SpotIMAPI.sso("<SECRET_JWT>")
    .then(response => {
        console.log(response);
      })
    .catch(error => {
        console.error(error);
      })
```

##### Logout
Call SpotIm logout API whenever a user logs out of your system

###### Example
```javascript
SpotIMAPI.logout();
```

##### Login status
An API to understand the status of the current SpotIm user.
Guest - Means this is a guest unregistered user. You should call startSSO/sooWithJWT if your own login status is 'user is logged in'
LoggedIn - Mean this is a registered user of SpotIm. You should avoid calling startSSO/sooWithJWT in this case. If you own status is 'user is logged out', you should SpotIm logout method

###### Example
```javascript
SpotIMAPI.getUserLoginStatus()
    .then(status => {
        console.log(status);
      })
    .catch(error => {
        console.error(error);
      })
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v1.8.0] - 2022-04-12
### Added
- Update Android SDK to version 1.9.8
- Update iOS SDK to version 1.11.0

### Fixed
- Custom dark mode background color is not applied

## [v1.7.0] - 2022-02-24
- Dark mode support

## [v1.6.0] - 2021-12-07

### Fixed
- Fix crash on RN 0.66

## [v1.5.1] - 2021-11-25

### Changed
- Depend on React-Core in iOS podspec

## [v1.5.0] - 2021-11-18

### Added
- Support Android 11 (API level 30)
- Update Android SDK v1.7.0
- Update iOS SDK v1.6.8

## [v1.4.4] - 2021-10-31

### Fixed
- iOS flag to open login screen on root VC issue with reply to comment
- Android flag to open login screen on root Activity issue with logged in user

### Changed
- Update Android SDK version to 1.6.3
- Update iOS SDK version to 1.6.7

## [v1.4.3] - 2021-10-26

### Added
- Update iOS SDK v1.6.6

### Fixed
- SpotIM component UIViewControllerHierarchyIncosistencyException issue

### Changed
- Sample app improvements - navigation and simulate login

## [v1.4.2] - 2021-10-21

### Fixed
- [Android] Login Screen "hiding" below "Create Comment" instead of on top - flag for open the login screen on the root Activity.

## [v1.4.1] - 2021-10-14

### Fixed
- [iOS] Login Screen "hiding" below "Create Comment" instead of on top - flag for open the login screen on the root VC.


## [v1.4.0] - 2021-10-10

### Fixed
- Analytics event duplications
- Memory leak when re-render SpotIm View
- SpotIm component should be update when changing postId prop

### Changed
- SDK initialization with SpotIMAPI module in root component constructor
- Remove spotId from SpotIm component props
- Android - no need to add android SDK dependency and to initialize the SDK from the native code

## [v1.3.0] - 2021-09-29

- Update iOS SDK v1.6.2
- Update Android SDK v1.6.1
- Analytics Event callback

## [v1.2.1] - 2021-09-12

- Update the react native iOS podspec file with iOS SDK v1.5.13
- Remove Google-Mobile-Ads-SDK dependency from podspec

## [v1.2.0] - 2021-05-27

- Update the react native iOS podspec file with iOS SDK v1.5.4 (move to Xcframework solution)
- Support for Xcode 14.5

## [v1.1.0] - 2021-02-10

- Update the react native iOS podspec file with iOS SDK v1.1.1
- `s.dependency "Google-Mobile-Ads-SDK", '~> 7.69.0'`



## [v1.0.14] - 2020-11-2
- Fixed crash when commiting the pre-conversation fragment after onSaveInstanceState called

## [v1.0.13] - 2020-10-21
### Fixed
- Android height issues
- Android crash when switching fast between screens that contain the SpotIm view
- iOS Crash when API returns an error

[Unreleased]: https://github.com/SpotIM/spotim-react-native-sdk/compare/v1.0.13...master
[1.0.13]: https://github.com/SpotIM/spotim-react-native-sdk/compare/1.0.12...v1.0.13


example/App.js
/**
 * Sample React Native App
 *
 * adapted from App.js generated by the following command:
 *
 * react-native init example
 *
 * https://github.com/facebook/react-native
 */

import { Platform, ScrollView, StyleSheet, Text, View, Button } from 'react-native';
import React, { Component } from 'react';
import type {Node} from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';
import 'react-native-gesture-handler';
import { SpotIM, SpotIMAPI, SpotIMEventEmitter } from '@spot.im/react-native-spotim';
import MainScreen from './Screens/MainScreen.js'
import ArticlesScreen from './Screens/ArticlesScreen.js'
import ArticleScreen from './Screens/ArticleScreen.js'
import LoginScreen from './Screens/LoginScreen.js'

const Stack = createStackNavigator();
export default class App extends Component {
  constructor(props) {
    super(props)
  }

  onTrackAnalyticsEvent(event) {
    console.log('\n', event.type, '\n', event);
  }

  componentDidMount() {
    // Show analytics events
    SpotIMEventEmitter.addListener('trackAnalyticsEvent', this.onTrackAnalyticsEvent);
  }

  render() {
    return (
      <NavigationContainer>
        <Stack.Navigator initialRouteName="Home">
          <Stack.Screen name="Home"component={MainScreen}/>
          <Stack.Screen name="Articles" component={ArticlesScreen} />
          <Stack.Screen name="Article" component={ArticleScreen} />
          <Stack.Screen name="Login" component={LoginScreen} />
        </Stack.Navigator>
      </NavigationContainer>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginVertical: 30,
  },
  welcome: {
    fontSize: 24,
    color: 'blue',
    textAlign: 'center',
    marginTop: 40,
    marginBottom: 20,
  },
});
